homeassistant/components/habitica/calendar.py

```diff
@@ -0,0 +1,261 @@
+            and not task["completed"]
+            and task["everyX"]
+            if (next_recurrence := build_rrule(task).after(today, inc=True))
+        ]
+        events_sorted = sorted(
+            events,
+            key=lambda event: (
+                event.start,
+                self.coordinator.data.user["tasksOrder"]["dailys"].index(event.uid),
+            ),
+        )
+
+        return events_sorted[0] if events_sorted else None
+
+    async def async_get_events(
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[CalendarEvent]:
+        """Return calendar events within a datetime range."""
+
+        today = to_date(self.coordinator.data.user["lastCron"])
+        # returns only todays and future dailies.
+        # If a daily is completed it will not be shown for today but still future recurrences
+        # If the cron hasn't run, not completed dailies are yesterdailies and displayed yesterday
+        return [
+            CalendarEvent(
+                start=start,
+                end=start,
+                summary=task["text"],
+                description=task["notes"],
+                uid=task["id"],
+                rrule=str(build_rrule(task))[30:],
+            )
+            for task in self.coordinator.data.tasks
+            if task["type"] == HabiticaTaskType.DAILY and task["everyX"]
+            for recurrence in build_rrule(task).between(start_date, end_date, inc=True)
+            if (start := recurrence.date()) > today
+            or (start == today and not task["completed"])
+        ]
+
+
+class HabiticaRemindersCalendarEntity(HabiticaCalendarEntity):
+    """Habitica reminders calendar entity."""
+
+    entity_description = CalendarEntityDescription(
+        key=HabiticaCalendar.REMINDERS,
+        translation_key=HabiticaCalendar.REMINDERS,
+    )
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the next upcoming event."""
+        events_sorted = None
+        now = dt_util.now()
+        events = [
+            CalendarEvent(
+                start=start,
+                end=start + timedelta(hours=1),
+                summary=task["text"],
+                description=task["notes"],
+                uid=f"{task["id"]}_{reminder["id"]}",
+            )
+            for task in self.coordinator.data.tasks
+            if task["type"] in (HabiticaTaskType.DAILY, HabiticaTaskType.TODO)
+            and not task["completed"]
+            for reminder in task.get("reminders", [])
+            if (
+                start := datetime.fromisoformat(reminder["time"]).replace(
+                    tzinfo=dt_util.DEFAULT_TIME_ZONE
+                )
+            )
+            >= (now - timedelta(hours=1))
+        ]
+
+        events_sorted = sorted(
+            events,
+            key=lambda event: event.start,
+        )
+
+        # return the last reminder that went off in the last hour
```

Comment:
This doesn't seem quite right to me since it is potentially hiding additional upcoming events? Current the rules are this returns either the current or next upcoming CalendarEvent.

I believe a simple way to implement "active after" is to iterate over the sorted event list and find the first event where the start time or end time are greater than the "now" time like here https://github.com/allenporter/ical/blob/d34a26a0af5bdd2a56c6601e7c35c6c012c9f0d8/ical/iter.py#L371 (used by local_calendar)

Reply:
I would recommend implementing this logic for the upcoming event and the event date range search once, and just make each type of event supply a sorted list of `CalendarEvent` and hand the rest with shared code.  Having the variance between them may introduce some bugs if we don't find all the subtle differences here so its easier to just get them right once and do them the same way.  (The logic can be the same whether or not its an all day event just by converting the dates to a timezone aware date when doing the range comparison)

Reply:
I think i still need some tweeking for the reminders calendar, I'm considering adding them in a follow-up PR

Reply:
It's not implementing the spec correctly.

Reply:
I see, you removed from the pr.